assorted updates

jeffreyhanson opened the pull request.

jeffreyhanson commented:
I've been working on some minor updates. @ricschuster, when you get a chance, could you please take a look and let me know what you think? In particular, here's a list of the main updates.

**New features**

- New `add_min_penalties_objective()` function to generate solutions that focus on minimizing the penalties -- as much as possible -- whilst ensuring that (i) the cost of the solution does not exceed a budget and (ii) all feature representation targets are met. This function is designed to aid with hierarchical multi-objective optimization. It can be now used instead of specifying a minimum set objective with zero costs and a linear constraint to specify the budget.
    - `R/add_min_penalties_objective.R`
    - `tests/testthat/test_add_min_penalties_objective.R`
    - `R/penalties.R`
    - `R/objectives.R`
    - `R/compile.R`
    - Also, do you have any ideas for a better name for this function?

**Minor improvements and bug fixes**

- Update `adjacency_matrix()`, `intersecting_units()`, `rij_matrix()`, `eval_target_coverage_summary()`, and `problem()` to improve processing speed.
    - `R/adjacency_matrix.R`
    -  `R/intersecting_units.R`
    - `R/rij_matrix.R`
    -  `R/eval_target_coverage_summary.R`
    - `R/problem.R`
    - `R/problem-deprecated.R`
    - `R/ConservationProblem-class.R`
    - `R/planning_unit_indices.R` 
    - `tests/testthat/planning_unit_indices.R` 
    - `R/planning_unit_solution_status.R`
    - `R/memory.R
    - `tests/testthat/test_memory.R`
    - `R/presolve_check.R`
- Update `intersecting_units()` to accommodate very large `sf::st_sf()` objects.
   - `R/intersecting_units.R`
- Update `rij_matrix()` so that it has an optional `idx` parameter which can be used to specify planning unit indices if they have been pre-computed.
   - `R/rij_matrix.R`
   - `tests/testthat/test_rij_matrix.R`
- Update `add_locked_in_constraints()` and `add_locked_out_constraints()` to throw error messages with standardized grammar.
    - `R/add_locked_in_constraints.R`
    - `R/add_locked_out_constraints.R`
- Update internal functions for processing `sf::st_sf()` object to improve processing speed.
    - `R/st_geometry_classes.R`
    - `tests/testthat/test_st_geometry_classes.R`
    - `R/is_spatial_extents_overlap.R`
- Update `add_cplex_solver()` error message for failing to initialize CPLEX environment.
     - `R/add_cplex_solver.R`
- Update error message for internal matrix conversion function.
    - `R/as_Matrix.R`
 
**Documentation updates**
- Update Package overview and Calibrating trade-offs vignettes with information on the `add_min_penalties_objective()` function.
    - `vignettes/calibrating_trade-offs_tutorial.Rmd`
    - `vignettes/package_overview.Rmd.Rmd`
- Update `add_max_utility_objective()` documentation to make it clear that the function is simply maximizing a weighted sum of the features.
    - `R/add_max_utility_objective.R`

ricschuster commented:
Thanks very much @jeffreyhanson!
I've started my review but am not finished yet.

One question I have already for function `add_min_penalties_objective` is whether "Minimize 0" does make sense and if it does should we explain in the Mathematical formulation why that is? As its presented now I'm pretty sure many users will be confused by that formulation.

jeffreyhanson commented:
Awesome - thanks! Yeah, that's a really good point. I'll push a commit in a moment to explain this.

ricschuster commented:
> Awesome - thanks! Yeah, that's a really good point. I'll push a commit in a moment to explain this.

Great! Thanks for the updated text.

ricschuster reviewed (approved):
Thanks very much for this update!
Everything that I looked at looks good.

As far as `add_min_penalties_objective` function name goes, I don't have any useful suggestions for a different name. What you have now makes sense to me.

jeffreyhanson commented:
Brilliant - thanks @ricschuster! Is there anything else you think we should put into this update? I could look into adding #337?

ricschuster commented:
I think that's a great idea if you are up for it.

jeffreyhanson commented:
Sure - I'll take a look and see what I can come up with.

jeffreyhanson commented:
I've implemented the rank importance method. @ricschuster, when you get a chance, could you please take a look and let me know what you think? In particular, I'd really appreciate it if you could read through the documentation for the function and see if you have any suggestions for improving it? I'd also appreciate it if you could double check it with some example data and seeing if you can break it to check that the error messages get thrown correctly? The updates are in the following files:

* Implementing the function and its unit tests.
    - `R/eval_rank_importance.R`
    - `tests/eval_rank_importance.R`
* Updating solver functions. These changes involve adding add new `x$set_constraint_rhs()` and `x$set_start_solution()` methods so that the rank importance method can quickly modify and solve an updated version of a problem. Although it would be easier/cleaner to add such methods to the `OptimizationProblem` class, the implemented changes will reduce memory requirements and run times for large problems.
    - `R/Solver-class.R`
    - `R/add_cbc_solver.R`
    - `R/add_cplex_solver.R`
    - `R/add_gurobi_solver.R`
    - `R/add_high_solvers.R`
    - `R/add_lpsymphony_solver.R`
    - `R/add_rsymphony_solver.R`
    - `tests/testthat/test_add_cbc_solver.R`
    - `tests/testthat/test_add_cplex_solver.R`
    - `tests/testthat/test_add_gurobi_solver.R`
    - `tests/testthat/test_add_high_solvers.R`
    - `tests/testthat/test_add_lpsymphony_solver.R`
    - `tests/testthat/test_add_rsymphony_solver.R`
* Update `eval_cost_summary()` function. This change is to move the main processing for this function into a new internal function (note that `eval_cost_summary()` still provides the same API/output/behavior). This way the new `eval_rank_importance()` function can avoid overhead associated with  calling `eval_cost_summary()` by calling the internal function directly.
  - `R/eval_cost_summary.R`
* Update documentation to refer to the new rank importance method. Since this method will probably become the new best method for moderate/large sized problems, I have updated the README and main tutorial vignette to use this method.
    - `NEWS.md`
    - `importances.R`
    - `README.Rmd`
    - `vignettes/prioritizr.Rmd`
    - `vignettes/package_overview.Rmd`
* Updates to improve test code coverage
    - `tests/testthat/test_is_spatial_extents_overlap.R`
* Minor documentation improvements and fixing typos
    - `R/ConservationProblem-class.R`
    - `R/eval_rare_richness_importance.R`: Detailed guidance for the function is moved to the Details section. 
    - `R/planning_unit_solution_status.R`: Fix typo in error message.
    - `inst/CITATION`: Fix typo in citation message.
    - `vignettes/gurobi_installation_guide.Rmd`: Fix broken URL.
    - `R/write_problem.R`: Fix broken URLs.

jeffreyhanson commented:
Since rcbc was failing to install for R devel, I had the prioritzr GitHub Actions configured to install a developmental version of rcbc with a patch for testing the prioritizr PR. The developmental version of  rcbc was just merged into the master branch, so I just updated the GitHub Actions for prioritizr to use the master branch version of rcbc. Just giving you a heads, the previous commit doesn't change much and I think the PR is still ready for review.

ricschuster commented:
Thanks very much Jeff!
Just back from 2 weeks of work travel and will try to get to this later in the week.

jeffreyhanson commented:
Sounds great - thanks!

ricschuster commented:
I've looked through the changes/additions now and everything looks good to me. I've only fixed one small typo I found.

I have only done a little bit of testing for generate errors, but so far everything works well.
Will try to get more testing done later in the week.

Overall, this looks great and great addition with the `eval_rank_importance` function!

jeffreyhanson commented:
Awesome - thank you very much @ricschuster! Yeah, I'd really appreciate thorough testing on this new function. It has a fair number of cases where certain inputs should produce errors (e.g., using problems with constraints that aren't locked in/out, phylogenetic diversity objective function without phylo data, specifying incorrectly formatted budgets for multi-zone problems, etc).

ricschuster commented:
Hi @jeffreyhanson,

Sorry for having taken so long to get back to you on this. I am having a really hard time finding room in my schedule to test this properly. What do you think about asking Sandra and maybe some others from the community to test this new function? Given my other committments I'm just not sure when I might get to this.
Sorry about that.

ricschuster commented:
Xavier might also be a good person to ask to test this.

ricschuster commented:
> Also, just to check, are there should be any additional bug fixes or features we want to bundle in this update? I don't think so, but I wanted to check? There might be a few more papers to add to the publication record, but that's all I can think of.

No more bug fixes or features that I can think of.

sandra-neubert commented:
Hi @jeffreyhanson and @ricschuster,

I have done the following testing (+ added comments where I have some) for the new `eval_rank_importance()` function.

The documentation was all clear to me, the only thing where some more information might be useful is for the plot of the objective values vs the rank.

I did all the testing with some of my own data (all `sf`). Let me know if you want me to do some `terra` specific testing.

General:

- Budgets: expected error thrown when not enough values specified or values are in the wrong format; Comment: I find the error message a little misleading because it gives the same message (One of `n` or `budgets` must be specified (not both)) when n AND budgets are specified or when neither of them is specified (but it sounds like then the problem would also be that both are specified). Could be changed to what you have in the parameter description: “Either `n` or `budgets` (not both) must be specified” (I also fixed a typo here) or “Exactly one of `n` or `budgets` must be specified”

Constraints:

- Tested locked-in constraints and linear constraints; throws expected error when no budget is specified in a problem with linear constraints

Objective functions:

- Tested a combination of objectives functions with different objectives in `eval_rank_importance()`. My main comment here is: I’d maybe add a warning or note when a different objective function is used than what’s in the original problem definition? Especially when the min set uses the default min shortfall  in `eval_rank_importance()` for calculating rank importance, but could be useful in general.

Zones:

- I’ve done some testing with zones already, and so far everything is working as expected, but can add some more - I can probably get to this next Wednesday.

Did you want me to test anything else in particular?

ricschuster commented:
@sandra-neubert just a quick general comment from me: thank you very much for looking into this! I very much appreciate your help with `prioritizr`.

jeffreyhanson commented:
Thanks so much @sandra-neubert! I really appreciate this. Also, thanks for catching those typos in the documentation!

Regarding your questions/suggestions:

> Budgets: expected error thrown when not enough values specified or values are in the wrong format; Comment: I find the error message a little misleading because it gives the same message (One of n or budgets must be specified (not both)) when n AND budgets are specified or when neither of them is specified (but it sounds like then the problem would also be that both are specified). Could be changed to what you have in the parameter description: “Either n or budgets (not both) must be specified” (I also fixed a typo here) or “Exactly one of n or budgets must be specified”

Yeah, thanks for raising this. It's really important to me that we have helpful error messages. The reason I originally had "One of n or budgets must be specified" was because I waned to follow similar terminology/phrasing to tidyverse error messages (e.g., "one of" appears in a lot of rlang's errors; https://github.com/search?q=repo%3Ar-lib%2Frlang%20%22one%20of%22&type=code). Looking at rlang again, however, I see that they use "Exactly one of ..." for cases like this. What do you think about that? If we went with "Exactly one of n or budgets (not both) must be specified.", how would that sound?

> Tested a combination of objectives functions with different objectives in eval_rank_importance(). My main comment here is: I’d maybe add a warning or note when a different objective function is used than what’s in the original problem definition? Especially when the min set uses the default min shortfall in eval_rank_importance() for calculating rank importance, but could be useful in general.

Yeah, I agree it's important to throw warnings when the software might not be doing something that intuitively makes sense or is making certain assumptions that might not be true. But I'm not sure that's needed here? In particular, the default behavior (except for min set problems) is to use the same objective function as the original problem -- so if a different objective function is used for calculating ranks, then this is something that the user has delibrately decided that they want. I don't think outputting warnings to simply re-iterate function arguments is particularly useful here? Also, I agree that it's useful for users to know that the ranks calculated for the min set problem default to using the min shortfall objective, but I don't think providing this as warning would make for a good user experience. So, I think having this explained in the documentation is ideal? Please let me know if you think isn't clear in the doucmentation though, or have any suggestions for making this clearer?

> I’ve done some testing with zones already, and so far everything is working as expected, but can add some more - I can probably get to this next Wednesday.

Awesome - thanks so much! Yeah, if you cold please test the zones functionality thoroughly - I'd really appreciate it. I think this probably where potential bugs might happen.

How does all that sound?

sandra-neubert commented:
No worries at all!

> Yeah, thanks for raising this. It's really important to me that we have helpful error messages. The reason I originally had "One of n or budgets must be specified" was because I waned to follow similar terminology/phrasing to tidyverse error messages (e.g., "one of" appears in a lot of rlang's errors; https://github.com/search?q=repo%3Ar-lib%2Frlang%20%22one%20of%22&type=code). Looking at rlang again, however, I see that they use "Exactly one of ..." for cases like this. What do you think about that? If we went with "Exactly one of n or budgets (not both) must be specified.", how would that sound?
> 
@jeffreyhanson, sounds good!

> Yeah, I agree it's important to throw warnings when the software might not be doing something that intuitively makes sense or is making certain assumptions that might not be true. But I'm not sure that's needed here? In particular, the default behavior (except for min set problems) is to use the same objective function as the original problem -- so if a different objective function is used for calculating ranks, then this is something that the user has delibrately decided that they want. I don't think outputting warnings to simply re-iterate function arguments is particularly useful here? Also, I agree that it's useful for users to know that the ranks calculated for the min set problem default to using the min shortfall objective, but I don't think providing this as warning would make for a good user experience. So, I think having this explained in the documentation is ideal? Please let me know if you think isn't clear in the doucmentation though, or have any suggestions for making this clearer?
>  I didn't mean having a warning in general for the default setting, only whenever the used objective does not match what hasn't been used in the problem. But I agree that it should be clear enough in the documentation.

I also did some more testing with zones now: setting up some problems with multiple different constraints (linear, locked-in, neighbor constraints), and also tried to think of ways the budgets could be set up the wrong way. It all made sense to me and I got errors when I expected them. The only thing that I was initially a bit confused by was that setting budgets in a single column (so not having a column per zone) worked for zones without me setting  "by_zone = FALSE". But maybe I just missed something in the documentation here.

Let me know if there's anything else that would be useful to test.

jeffreyhanson commented:
Thanks so much! Yeah, `by_zone` only has an effect when `n` is supplied. If `budgets` is specified instead, then `by_zone` has no effect on the process. I'll update the documentation to make this clearer, and have the function throw a warning if `by_zone` is the non-default value and `budgets` is supplied. How does that sound?

sandra-neubert commented:
Sounds good @jeffreyhanson!

jeffreyhanson commented:
Awesome - thanks!

jeffreyhanson commented:
@ricschuster, I've just rebuilt the website and tweaked the cran-comments. Is there anything else needed before we merge?

ricschuster commented:
@jeffreyhanson thanks very much! Nothing else comes to mind other than any addtitional testing @sandra-neubert would like to do. If there is nothing else to test, I'd say we are good to merge.

jeffreyhanson commented:
Ok great! I've just checked the revdeps (i.e., check compatibility with packages on CRAN that depend on prioritizr). @ricschuster, could you do a final winbuilder check and check that you're happy with the CRAN comments? If that's all good, I'll merge the branch and then could you submit to CRAN?

ricschuster commented:
Thanks very much @jeffreyhanson!
winbuilder checks are good too so ready to merge and after that I can submit to CRAN.

jeffreyhanson commented:
Ah, just found some typos in the examples for min penalties obj. Sorry - I'll fix those and update the docs
